Tracking from everywhere

nzoba opened the pull request:
This PR contains:

- Tracking from everywhere (notification, library, manga_details and reader) when chapter marked as read/unread (active via settings, default is only reader like before), with possibility to decrease the tracking chapter number except for reader.
Not sure about this one since it could increase tracking traffic a bit too much, it's up to you. I added a 3s delay before updating tracker from manga_details to limit that, this delay is reset if you toggle a new chapter that should update the tracker but is it enough ?
- Paused tracking (by long click on tracking icon in manga_details of a tracked manga or via settings), probably useful only if tracking from everywhere is used


| Tracking settings | Paused tracking |
|----|----|
| ![Tracking settings](https://user-images.githubusercontent.com/55888232/155899880-036a7884-48b5-4af0-831b-ee23d40c45eb.png) | ![Paused tracking](https://user-images.githubusercontent.com/55888232/155900905-d24242d0-d3d5-43e4-b057-d0ce430e6845.png) |

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaDetailsPresenter.kt:
```diff
@@ -487,10 +494,63 @@ class MangaDetailsPresenter(
+
+        // We want these to execute even if the presenter is destroyed so launch on GlobalScope
+        job?.cancel()
+        job = launchIO {
```
as mentioned it's fine to use presenterscope for this, it's short calls

Jays2Kings replied:
actually, thinking again of the person who will close pages the moment they mark something as read, I makes sense to useon globalscope. but i'd still use something else like
`(controller.activity as AppCompatActivity).lifecycleScope.launchIO`

Jays2Kings reviewed (commented):
I've been thinking on this one for a while but I dont think I'm a fan of either feature for this.

The tracking everywhere could be simplified to a "track when marked as read" setting (like how theres a setting for deleting downloads when marked, seperate from the delete after actually reading) instead of having it be so robust for each way to mark chapters. You can then make a helper function to update tracking since it's 4 (really 3) places that should have the logic: 
library mark as read (skip this one), recents, and manga details all have a snackbar that has to dismiss before deleting downloads, you can tie updating tracking to that.
The 3rd/4th is notifications and you have that one already

As for the paused tracking, I think its better to make a setting that pauses tracking while incognito mode is on. I can't think of any other time someone would want to pause tracking but not read status

nzoba commented:
So I added the changes you requested :
- Changed tracking everywhere to track when marked as read. You are right it's far better and totally unnecessary to differentiate each case. I mutualized as much code as possible and removed library tracking.
- Removed the paused tracking setting. At first I added it in case of a desynchro between tracker last read chapter and tachi last read, the possibility to pause tracking was helpful to avoid updating tracker when doing a manual resynchro from tachi. But it's not necessary anymore since I added the following point
- Added a button to the trackers holder to update chapters progress from tracker (particularly helpful if you are reading on multiple devices or when adding a new manga already started). It could probably be improved by allowing to define a default tracker and do something like this : if the tracker last read chapter is bigger than tachi last read -> automatically update tachi last read.
But for now I did a basic button and probably won't implement this idea.

nzoba commented on app/src/main/java/eu/kanade/tachiyomi/data/notification/NotificationReceiver.kt:
```diff
@@ -192,18 +201,76 @@ class NotificationReceiver : BroadcastReceiver() {
+                            trackings.add("$mangaId:${track.sync_id}:$newCountChapter")
+                            preferences.trackingsToAddOnline().set(trackings)
+                            DelayedTrackingUpdateJob.setupTask(preferences.context)
+                        } else {
```
I mutualized this method code between all cases so I didn't remove it but if you prefer I can change it

Jays2Kings reviewed (commented):
Lots of comments again 😅, but they're more for assistance than heavy changes

nzoba commented on app/src/main/res/values/strings.xml:
```diff
@@ -559,6 +559,7 @@
     <string name="enhanced_services">Enhanced services</string>
     <string name="enhanced_tracking_info">Services that provide enhanced features for specific sources. Manga are automatically tracked when added to your library.</string>
     <string name="sync_chapters_after_reading">Sync chapters after reading</string>
+    <string name="sync_chapters_marked_read">Sync chapters when marked as read</string>
```
Ok. Should I change the one above it to `update_tracking_after_reading` ? If so what should I do about the existing translations ?

Jays2Kings replied:
Yeah that can be made to match upstream too. But only update the English/main strings, don't touch the other languages keys, they'll be updated and cleaned up automatically

nzoba commented on app/src/main/res/drawable/ic_cloud_download_24dp.xml:
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Was added on purpose for the update chapters progress from tracker feature but since you don't want it in this PR, I'll remove it

nzoba commented on app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaDetailsPresenter.kt:
```diff
@@ -931,6 +942,23 @@ class MangaDetailsPresenter(
         }
     }
 
+    fun updateChaptersProgress(lastChapterRead: Int) {
```
Okay, will do

nzoba commented on app/src/main/java/eu/kanade/tachiyomi/util/chapter/ChapterTrackSync.kt:
```diff
@@ -40,3 +53,72 @@ fun syncChaptersWithTrackServiceTwoWay(db: DatabaseHelper, chapters: List<Chapte
+ * Starts the service that updates the last chapter read in sync services. This operation
+ * will run in a background thread and errors are ignored.
+ */
+fun Controller.updateTrackChapterMarkedAsRead(
+    db: DatabaseHelper,
+    preferences: PreferencesHelper,
+    oldLastChapter: Chapter?,
+    newLastChapter: Chapter?,
+    mangaId: Long?,
+    delay: Long = 3000
```
What is the purpose of the callback here ?

Jays2Kings replied:
Honestly I forgot the full reason by now lol. I think it was to update trackers in manga details page, since that page shows tracking it needs to be updated after it finishes calling the sevices

nzoba commented on app/src/main/java/eu/kanade/tachiyomi/util/chapter/ChapterTrackSync.kt:
```diff
@@ -40,3 +53,72 @@ fun syncChaptersWithTrackServiceTwoWay(db: DatabaseHelper, chapters: List<Chapte
+    val trackList = db.getTracks(mangaId).executeAsBlocking()
+    trackList.map { track ->
+        val service = trackManager.getService(track.sync_id)
+        if (service != null && service.isLogged) {
+            val shouldCustomCount = listOf(
+                abs(track.last_chapter_read - oldChapterRead), oldChapterRead, track.last_chapter_read
+            ).all { it > 15 }
+            val newCountChapter = if (shouldCustomCount) {
+                (track.last_chapter_read + (newChapterRead - oldChapterRead)).coerceAtLeast(0)
+            } else newChapterRead
+            if (!preferences.context.isOnline()) {
+                val trackings = preferences.trackingsToAddOnline().get().toMutableSet()
+                val currentTracking = trackings.find { it.startsWith("$mangaId:${track.sync_id}:") }
+                trackings.remove(currentTracking)
+                trackings.add("$mangaId:${track.sync_id}:$newCountChapter")
+                preferences.trackingsToAddOnline().set(trackings)
+                DelayedTrackingUpdateJob.setupTask(preferences.context)
+            } else {
+                try {
+                    track.last_chapter_read = newCountChapter
+                    service.update(track, true)
+                    db.insertTrack(track).executeAsBlocking()
+                } catch (e: Exception) {
+                    Timber.e(e)
+                }
+            }
```
Well this is almost like how it was before my changes. There is two reasons why I added the old chapter (as far as I remember, did that a few months ago): 
- Allow to decrease last chapter number (except for reader)
- Allow correct tracking for cases like seasons. For example if a manga source has all chapters but the tracker split it into seasons then the tracked value would be incorrect.
Hence this piece of code (the 15 value is random, I thought that a difference of more than 15 chapters between tracker last read and app last read would mean either a season or purposefully different), to increase the tracker chapter by the same difference between old and new app chapters :
```
val shouldCustomCount = listOf(
                abs(track.last_chapter_read - oldChapterRead), oldChapterRead, track.last_chapter_read
            ).all { it > 15 }
            val newCountChapter = if (shouldCustomCount) {
                (track.last_chapter_read + (newChapterRead - oldChapterRead)).coerceAtLeast(0)
            } else newChapterRead
```

Jays2Kings replied:
I think it's overkill to support decrease in chapters read, since it's rarely a case where users are going backwards in progress
For point 2 I'm not sure if you mean the source has all the chapters combined into seasons or the tracker lists the chapters as seasons. If it's the latter that tracker needs to be reported for incorrect formatting. Just about every tracking service supports volumes (seasons) as a separate value, it's just not something that's updated on tachi.

So yeah I think going back to the basics is overall better

nzoba replied:
Yeah the chapter's decrease might not be necessary
About point 2 here is a concrete example: https://anilist.co/manga/129705/Chakage-Salja-Season-1/
Anilist decided to split this manhwa between seasons (why ? no idea), so if a source list all the chapters at once, the last read chapter will be wrong. Like chapter 67 in tachi should be 16 in anilist season 2 (with season 1 marked as completed), but the current feature would force the tracker to put 67 on season 2
I encountered this case a few times (sometimes reversed) but not sure how common it is

Jays2Kings replied:
Well thats annoying, but honestly I'd just leave this be, and tackle it another day. This pr is really clean right now as is so I'd prefer leaving it out for now

nzoba replied:
Okay, fine with me

nzoba commented:
I think it should be good now (except if you see something else)
I removed the old chapter parameter even though I think it would be good to allow custom tracking value like I explained
I could not test the notification receiver since I did not receive new chapters but it seems good to me

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaDetailsPresenter.kt:
```diff
@@ -98,6 +100,7 @@ class MangaDetailsPresenter(
     private var tracks = emptyList<Track>()
 
     var trackList: List<TrackItem> = emptyList()
+    private var job: Job? = null
```
not used

Jays2Kings commented on app/src/main/java/eu/kanade/tachiyomi/util/chapter/ChapterTrackSync.kt:
```diff
@@ -40,3 +48,68 @@ fun syncChaptersWithTrackServiceTwoWay(db: DatabaseHelper, chapters: List<Chapte
+    preferences: PreferencesHelper,
+    newLastChapter: Chapter?,
+    mangaId: Long?,
+    fetchTracks: (suspend () -> Unit)? = null,
+    delay: Long = 0
```
Id switch these around, and make delays default 3000, since only notification is using 0. Flipping them around also allows updateTrackChapterMarkedAsRead(...) { fetchtracks }
```suggestion
    delay: Long = 3000,
    fetchTracks: (suspend () -> Unit)? = null,
```

nzoba replied:
Done

Jays2Kings reviewed (approved):
🚀
